go-controller/pkg/ovn/zone_interconnect/zone_ic_handler.go

```diff
@@ -0,0 +1,665 @@
+		if !foundNodes.Has(lportNode) {
+			staleNodes = append(staleNodes, lportNode)
+		}
+	}
+
+	for _, staleNode := range staleNodes {
+		if err := zic.cleanupNode(staleNode); err != nil {
+			klog.Errorf("Failed to cleanup the interconnect resources from OVN Northbound db for the stale node %s : %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Cleanup deletes the transit switch for the network
+func (zic *ZoneInterconnectHandler) Cleanup() error {
+	klog.Infof("Deleting the transit switch [%s] for the network [%s]", zic.networkTransitSwitchName, zic.GetNetworkName())
+	return libovsdbops.DeleteLogicalSwitch(zic.nbClient, zic.networkTransitSwitchName)
+}
+
+// createLocalZoneNodeResources creates the local zone node resources for interconnect
+//   - creates Transit switch if it doesn't yet exit
+//   - creates a logical switch port of type "router" in the transit switch with the name as - <network_name>.tstor-<node_name>
+//     Eg. if the node name is ovn-worker and the network is default, the name would be - tstor-ovn-worker
+//     if the node name is ovn-worker and the network name is blue, the logical port name would be - blue.tstor-ovn-worker
+//   - creates a logical router port in the ovn_cluster_router with the name - <network_name>.rtots-<node_name> and connects
+//     to the node logical switch port in the transit switch
+//   - remove any stale static routes in the ovn_cluster_router for the node
+func (zic *ZoneInterconnectHandler) createLocalZoneNodeResources(node *corev1.Node, nodeID int) error {
+	nodeTransitSwitchPortIPs, err := util.ParseNodeTransitSwitchPortAddrs(node)
+	if err != nil || len(nodeTransitSwitchPortIPs) == 0 {
+		return fmt.Errorf("failed to get the node transit switch port ips for node %s: %w", node.Name, err)
+	}
+
+	networkId, err := util.ParseNetworkIDAnnotation(node, zic.GetNetworkName())
+	if err != nil {
+		return fmt.Errorf("failed to get the network id for the network %s on node %s: %v", zic.GetNetworkName(), node.Name, err)
+	}
+
+	transitRouterPortMac := util.IPAddrToHWAddr(nodeTransitSwitchPortIPs[0].IP)
+	var transitRouterPortNetworks []string
+	for _, ip := range nodeTransitSwitchPortIPs {
+		transitRouterPortNetworks = append(transitRouterPortNetworks, ip.String())
+	}
+
+	ts := &nbdb.LogicalSwitch{
+		Name: zic.networkTransitSwitchName,
+	}
+
+	_, err = libovsdbops.GetLogicalSwitch(zic.nbClient, ts)
+	if err != nil {
+		if errors.Is(err, libovsdbclient.ErrNotFound) {
+			transitSwitchTunnelKey := BaseTransitSwitchTunnelKey + networkId
+			// Create transit switch
+			ts.OtherConfig = map[string]string{
+				"interconn-ts":             zic.networkTransitSwitchName,
+				"requested-tnl-key":        strconv.Itoa(transitSwitchTunnelKey),
+				"mcast_snoop":              "true",
+				"mcast_flood_unregistered": "true",
+			}
+			if err := libovsdbops.CreateOrUpdateLogicalSwitch(zic.nbClient, ts); err != nil {
+				return fmt.Errorf("failed to create/update logical transit switch: %w", err)
+			}
```

Comment:
Any reason you are doing the Get before the CreateOrUpdate? You are not reconciling the switch if you don't check the data it has is actually the one you want to set. We usually just do the CreateOrUpdate directly.

Reply:
This comment is still outstanding.